### DISCUSSION TITLE
Virtual element

### DISCUSSION THREAD

**Discussion by msayyafuofa** [idaholab/moose#20462] [Q&A General] [answered]
Hi, 

I am a newbie here. Over the past two weeks, I have been going through the tutorials and workshops and have been fascinated with this framework and the materials available for learning. 

I have a general question. I am doing research on subsurface flow problems and occasionally dealing with systems with nonconvex meshes. If I am not mistaken, FEM is not the best method for general polygonal elements such as might be found in reservoir simulation. I was wondering if there is any plan in future to incorporate Virtual Element (VEM) in Moose. 

Thank you :-)

**Comment by hugary1995** [comment, marked as answer]
AFAIK MOOSE doesn't support Ansatz shape functions and polygonal elements yet. I think you'll have to start from libMesh if you want to develop it yourself, but I'm afraid that many low level operations assume convex elements.

**Reply by msayyafuofa** [reply]
Thank you for the clarification. Is there any plan/interest for the implementation of VEM in the near future (in a year or so?)?

**Reply by hugary1995** [reply]
I don't know to be honest. Perhaps you should unmark this as the answer, and others can see and comment on your question.

**Reply by loganharbour** [reply]
Ping @roystgnr

**Reply by roystgnr** [reply]
There's been some interest in adding support for more general polygonal elements (well, hexagons, but if we were to do the work anyway I'd hope to handle general N) to libMesh, but just using macroelement subdivision of convex polygons, not a fancy new weak form, and even for the macroelement implementation I'm not sure if there's any funding/time for it in the next year or two.

The first Google hit I find for "Virtual Element Method" is someone in the UK who likes the blue Beamer LaTeX theme and the cross benchmark problem for Cahn-Hilliard, so I can already tell we could have been best friends in another life, but I'm afraid that here and now this is the first I've heard of VEM.

**Reply by roystgnr** [reply]
Oh, and to be clear, this probably is a good answer:

> AFAIK MOOSE doesn't support Ansatz shape functions and polygonal elements yet. I think you'll have to start from libMesh if you want to develop it yourself

Correct and correct.

> but I'm afraid that many low level operations assume convex elements

I think this is only correct for some particular physics implementations (contact in Moose?) not for anything at the libMesh level ... however libMesh does assume that it makes sense to talk about function definitions in "master" space cleanly distinct from the transformation between that space and "physical" space, and we can get some ugly code (see the Clough-Tocher shape functions for example) in cases where that's not the most sensible description.
